drivers: clock: rcar: Rework structure & filetree V2 (using a driver per SoC)

aaillet opened the pull request:
Hi,

This pull request is following the conversation I had in PR #43059 with @julien-massot and @gmarull.

Goals of this work are still to:
  - Deploy a clean clocks definitions file tree as close as Linux one as possible.
  - Use same clock names as linux ones for unification purpose.
  - To deploy a file tree and rework the drive so that it will be easier to add new Renesas SoCs later.

This work proposes an inverted way to rework the clock driver for Renesas R-Car platforms:
  - Not writing soc specific methods in the soc directory as I did before.
  - Deploying a driver per soc alongside a common driver and its header.
  - Enabling the right driver depending on used SoC.
  - Each SoC has it's own "compatible" field such as in Linux.
  - Clocks definition are the same as in Linux.
  - Files names are as close as Linux as possible.
  - It ensure a compatibility with Linux device trees on Clock controller fields.

The idea is to compare both versions and to choose which one is the best one to be merged in Zephyr.

These changes have been tested on r8a77951 SoC's powered H3ULCB board.

Signed-off-by: Aymeric Aillet [aymeric.aillet@iot.bzh](mailto:aymeric.aillet@iot.bzh)

aaillet commented:
Added covering for new clock driver files in MAINTAINER file.

aaillet commented:
Resolved CI issues by adding new files to CODEOWNER file.

henrikbrixandersen commented:
@aaillet Which one is it? This one or #43059?

aaillet commented:
> 

Both are implementing the same logic but not in the same way.
The first one #43059 was implementing a single driver and soc specific methods in the /soc folder while this one is deploying a driver per soc alongside a common driver in /drivers folder.
This second try has been done after a question @gmarull asked on the first version's PR.

aaillet commented:
> @aaillet Which one is it? This one or #43059?

Both are implementing the same logic but not in the same way.
The first one https://github.com/zephyrproject-rtos/zephyr/pull/43059 was implementing a single driver and soc specific methods in the /soc folder while this one is deploying a driver per soc alongside a common driver in /drivers folder.
This second try has been done after a question @gmarull asked on the first version's PR.

henrikbrixandersen commented:
@aaillet Without having reviewed this in detail, I much prefer this PR over #43059. I think you should decide which one, you would like reviewers to concentrate on and close the other one for now. Having two PRs solving the same issue is confusing.

aaillet commented:
> @aaillet Without having reviewed this in detail, I much prefer this PR over #43059. I think you should decide which one, you would like reviewers to concentrate on and close the other one for now. Having two PRs solving the same issue is confusing.

Jus did it, thank you for your advice.

julien-massot commented on drivers/clock_control/renesas/renesas-cpg-mssr.h:
```diff
@@ -0,0 +1,50 @@
+
+#ifdef CONFIG_SOC_SERIES_RCAR_GEN3
+/* Software Reset Clearing Register offsets */
+    #define SRSTCLR(i)      (0x940 + (i) * 4)
```
```suggestion
#define SRSTCLR(i)      (0x940 + (i) * 4)
```

henrikbrixandersen replied:
Please fix the indentation.

aaillet replied:
Fixed !

julien-massot commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -48,3 +47,8 @@ endif()
 zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_AGILEX clock_agilex_ll.c)
 zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_AGILEX clock_agilex.c)
+
+if(CONFIG_CLOCK_CONTROL_RCAR_CPG_MSSR)
```
perhaps we may create a new CMakeLists.txt in the renesas subfolder

henrikbrixandersen replied:
Yes, please create a separate CMakeList.txt file for each subdirectory. But is a new subdirectory really needed?

aaillet replied:
The new sub directory could be a nice addition in my opinion if it was done for each platform.
In the current state, the renesas sub directory is indeed not really needed, at least from a functional point of view.

aaillet replied:
I chose to move back drivers and headers into the common clock_control drivers directory.
If a new file tree is to be chosen for clarity, we would do it for all providers at the same time.

henrikbrixandersen commented on drivers/clock_control/renesas/r8a77951-cpg-mssr.c:
```diff
@@ -0,0 +1,152 @@
+/*
```
The convention is to use `_` as separator in file names - not `-`.

aaillet replied:
I used these file name to match Linux file names, are we not trying to be as close as possible from Linux code base if possible ?
I'll do the change then, thanks for your message.

aaillet replied:
renamed all files !

henrikbrixandersen commented on drivers/clock_control/renesas/r8a77951-cpg-mssr.c:
```diff
@@ -0,0 +1,152 @@
+
+#include <errno.h>
+#include <soc.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/renesas-cpg-mssr.h>
+#include <dt-bindings/clock/renesas/renesas-cpg-mssr.h>
+#include <dt-bindings/clock/renesas/r8a7795-cpg-mssr.h>
```
```suggestion
#include <zephyr/drivers/clock_control.h>
#include <zephyr/drivers/clock_control/renesas-cpg-mssr.h>
#include <zephyr/dt-bindings/clock/renesas/renesas-cpg-mssr.h>
#include <zephyr/dt-bindings/clock/renesas/r8a7795-cpg-mssr.h>
```

aaillet replied:
Thanks, forgot to migrate to the new prefix !

aaillet replied:
Migrated to new prefixes !

henrikbrixandersen commented on drivers/clock_control/renesas/r8a77951-cpg-mssr.c:
```diff
@@ -0,0 +1,152 @@
+#include "renesas-cpg-mssr.h"
+
+struct r8a77951_mssr_config {
+	uint32_t base_address;
```
```suggestion
	mm_reg_t base_address;
```

henrikbrixandersen commented on drivers/clock_control/renesas/r8a77951-cpg-mssr.c:
```diff
@@ -0,0 +1,152 @@
+
+	if (enable) {
+		if ((CANFDCKCR_PARENT_CLK_RATE % rate) != 0) {
+			__ASSERT(true, "Can not generate "
+				 "%u from CANFD parent clock", rate);
```
This looks wrong. Perhaps use `LOG_ERR()`?

aaillet replied:
Moved to LOG_ERR() and registered a new "log module".
Does it sound ok for you like that ?

henrikbrixandersen commented on drivers/clock_control/renesas/r8a77951-cpg-mssr.c:
```diff
@@ -0,0 +1,152 @@
+			ret = -EINVAL;
+			goto unlock;
+		}
+		divider = (CANFDCKCR_PARENT_CLK_RATE / rate) - 1;
```
Missing guard against `rate == 0`.

aaillet replied:
Added a guard against rate equal or less to 0.

henrikbrixandersen commented on drivers/clock_control/renesas/r8a77951-cpg-mssr.c:
```diff
@@ -0,0 +1,152 @@
+
+static int r8a77951_cpg_mssr_init(const struct device *dev)
+{
+	return 0;
```
```suggestion
	ARG_UNUSED(dev);

	return 0;
```

aaillet replied:
Added ARG_UNUSED(dev)

aaillet committed:
dts: rcar: Rework clock definition

This commit is reworking clock definition in
order to match linux filetree and definition
as much as possible.

- dt-bindings: rework renesas clocks dt-bindings
  - regroup renesas related dt-bindings in a folder
  - rename renesas rcar common dt-binding to match linux name
  - add soc specific dt-binding matching linux name
  - soc dt-bindings are defining clocks matching linux names

- dts: use new clocks names
  - move clocks definitions in SoC layer for each core clock entry

- driver: use new clocks names

As seen in this commit, we are declaring clocks for "R8A7795" SoC
to match linux names.
Linux is not declaring "R8A77951" SoC specific files because
its also supporting the first H3 SoC version numbered "R8A77950".

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
drivers: clock: rcar: Deploy a driver for each soc

This commit rewrite renesas R-Car clock driver in order
to be able to support any new SoC easier.

This work is so creating a clock driver per soc alongside a
common driver for all reneasas r-car boars.

- drivers: create a driver per soc
  - create a common driver
  - create a common header used by soc & common driver
  - create a soc specific driver calling for common driver

- dts: use new compatible
  - use old yaml as common yaml
  - create a new "child" yaml to define the new compatible field
  - change compatible in device tree

As in Linux, the driver can support both r8a77951 and r8a77950
SoC's so we decided to name the new driver as in Linux with Zephyr
prefix : "clock_control_r8a7795_cpg_mssr.c".

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
drivers: clock: rcar: Rename global includes file

Rename r-car clock driver global include file
in order to match other files names.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

mbolivar-nordic reviewed (dismissed):
@aaillet I'll take your word for it that these devicetree changes better align our bindings with Linux. No objections from me to the DT portions, and thanks for the cleanup. This approval is for the DT part only; the driver portions I leave to @henrikbrixandersen

henrikbrixandersen commented:
@aaillet Please resolve the merge conflicts.

aaillet commented:
> @aaillet I'll take your word for it that these devicetree changes better align our bindings with Linux. No objections from me to the DT portions, and thanks for the cleanup. This approval is for the DT part only; the driver portions I leave to @henrikbrixandersen

Hey @mbolivar-nordic , Henrik just validated main changes of this PR after some fixes.
I didn't touched DT related files with my updates but don't hesitate to check these files if you want.
Please re-approve it when you have some time, Thanks a lot.
